File: Homework1/Hw1/Parser.cs
namespace Hw1;

public static class Parser
{
    public static void ParseCalcArguments(string[] args, 
        out double val1, 
        out CalculatorOperation operation, 
        out double val2)
    {
        if (!IsArgLengthSupported(args)) throw new ArgumentException("You have to pass 3 arguments");
        
        if (!double.TryParse(args[0], out val1)) throw new ArgumentException("The first argument could not be converted to the double type");
        if (!double.TryParse(args[2], out val2)) throw new ArgumentException("The second argument could not be converted to the double type");
        operation = ParseOperation(args[1]);
    }

    private static bool IsArgLengthSupported(string[] args) => args.Length == 3;

    private static CalculatorOperation ParseOperation(string arg)
    {
        return arg switch
        {
            "+" => CalculatorOperation.Plus,
            "-" => CalculatorOperation.Minus ,
            "*" => CalculatorOperation.Multiply,
            "/" => CalculatorOperation.Divide,
            _ => throw new InvalidOperationException("Unknown operation")
        };
    }
}

File: Homework4/Hw4/Parser.fs
module Hw4.Parser

open System
open Hw4.Calculator


type CalcOptions = {
    arg1: float
    arg2: float
    operation: CalculatorOperation
}

let isArgLengthSupported (args : string[]) =
    args.Length = 3

let parseOperation (arg : string) =
    match arg with
        | "+" -> CalculatorOperation.Plus
        | "-" -> CalculatorOperation.Minus
        | "*" -> CalculatorOperation.Multiply
        | "/" -> CalculatorOperation.Divide
        | _ -> raise(ArgumentException "Unknown operation")
        
let parseArgument(str: string)(argNumb:string)  =
    match Double.TryParse str with
    | true,arg -> arg
    | _ -> raise(ArgumentException(argNumb+" argument could not be converted to the double type"))
    
let parseCalcArguments(args : string[]) =
    if not (isArgLengthSupported args) then raise(ArgumentException "You have to pass 3 arguments")

    { arg1 = parseArgument args[0] "First"
      arg2 = parseArgument args[2] "Second"
      operation = parseOperation args[1] }
    
    

    
    
    
    

